v2/blockstore/readwrite.go

```diff
@@ -227,30 +237,70 @@ func (b *ReadWrite) Discard() {
 
 // Finalize finalizes this blockstore by writing the CARv2 header, along with flattened index
 // for more efficient subsequent read.
+// This is the equivalent to calling FinalizeReadOnly and Close.
 // After this call, the blockstore can no longer be used.
 func (b *ReadWrite) Finalize() error {
+	b.ronly.mu.Lock()
+	defer b.ronly.mu.Unlock()
+
+	if err := b.finalizeReadOnlyWithoutMutex(); err != nil {
+		return err
+	}
+	if err := b.closeWithoutMutex(); err != nil {
+		return err
+	}
```

Comment:
```suggestion
	for _, err := range []error{b.finalizeReadOnlyWithoutMutex(), b.closeWithoutMutex()} {
		if err != nil {
			return err
		}
	}
```

This is a change of behaviour over the existing `Finalize()` but I think it might best to at least attempt to getting closer to closing the filehandle even if we error on the way there, since we're responsible for it, leaving the user with no ability to perform any cleanup themselves.

Sorry for the late suggestion, I was just doing [something similar](https://github.com/filecoin-project/lassie/pull/124/files#diff-f20cfdf182d4cab5bbd42b104051dd0e6ff93c4482fa9c56266db9107a1fea4a) recently and this came up.

What do you think?

Reply:
It still looks to me like trying to catch a falling knife, but why not, you are the best to judge what make sense.

Reply:
haha, yeah, maybe so, but an error completing the index really shouldn't result in a file descriptor left dangling if we can help it